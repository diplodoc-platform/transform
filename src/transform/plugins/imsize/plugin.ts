import type MarkdownIt from 'markdown-it';
import type ParserInline from 'markdown-it/lib/parser_inline';
import type Token from 'markdown-it/lib/token';

import {ImsizeAttr} from './const';
import {parseImageSize} from './helpers';

export type ImsizeOptions = {
    enableInlineStyling?: boolean;
};

export const imageWithSize = (md: MarkdownIt, opts?: ImsizeOptions): ParserInline.RuleInline => {
    // eslint-disable-next-line complexity
    return (state, silent) => {
        if (state.src.charCodeAt(state.pos) !== 0x21 /* ! */) {
            return false;
        }

        if (state.src.charCodeAt(state.pos + 1) !== 0x5b /* [ */) {
            return false;
        }

        const labelStart = state.pos + 2;
        const labelEnd = md.helpers.parseLinkLabel(state, state.pos + 1, false);

        // parser failed to find ']', so it's not a valid link
        if (labelEnd < 0) {
            return false;
        }

        let href = '';
        let title = '';
        let width = '';
        let height = '';

        const oldPos = state.pos;
        const max = state.posMax;

        let pos = labelEnd + 1;
        if (pos < max && state.src.charCodeAt(pos) === 0x28 /* ( */) {
            //
            // Inline link
            //

            // [link](  <href>  "title"  )
            //        ^^ skipping these spaces
            pos++;
            for (; pos < max; pos++) {
                const code = state.src.charCodeAt(pos);
                if (code !== 0x20 && code !== 0x0a) {
                    break;
                }
            }
            if (pos >= max) {
                return false;
            }

            // [link](  <href>  "title"  )
            //          ^^^^^^ parsing link destination
            let start = pos;
            const destResult = md.helpers.parseLinkDestination(state.src, pos, state.posMax);
            if (destResult.ok) {
                href = state.md.normalizeLink(destResult.str);
                if (state.md.validateLink(href)) {
                    pos = destResult.pos;
                } else {
                    href = '';
                }
            }

            // [link](  <href>  "title"  )
            //                ^^ skipping these spaces
            start = pos;
            for (; pos < max; pos++) {
                const code = state.src.charCodeAt(pos);
                if (code !== 0x20 && code !== 0x0a) {
                    break;
                }
            }

            // [link](  <href>  "title"  )
            //                  ^^^^^^^ parsing link title
            const titleResult = md.helpers.parseLinkTitle(state.src, pos, state.posMax);
            if (pos < max && start !== pos && titleResult.ok) {
                title = titleResult.str;
                pos = titleResult.pos;

                // [link](  <href>  "title"  )
                //                         ^^ skipping these spaces
                for (; pos < max; pos++) {
                    const code = state.src.charCodeAt(pos);
                    if (code !== 0x20 && code !== 0x0a) {
                        break;
                    }
                }
            } else {
                title = '';
            }

            // [link](  <href>  "title" =WxH  )
            //                          ^^^^ parsing image size
            if (pos - 1 >= 0) {
                const code = state.src.charCodeAt(pos - 1);

                // there must be at least one white spaces
                // between previous field and the size
                if (code === 0x20) {
                    const sizeResult = parseImageSize(state.src, pos, state.posMax);
                    if (sizeResult.ok) {
                        width = sizeResult.width;
                        height = sizeResult.height;
                        pos = sizeResult.pos;

                        // [link](  <href>  "title" =WxH  )
                        //                              ^^ skipping these spaces
                        for (; pos < max; pos++) {
                            const code = state.src.charCodeAt(pos);
                            // eslint-disable-next-line max-depth
                            if (code !== 0x20 && code !== 0x0a) {
                                break;
                            }
                        }
                    }
                }
            }

            if (pos >= max || state.src.charCodeAt(pos) !== 0x29 /* ) */) {
                state.pos = oldPos;
                return false;
            }
            pos++;
        } else {
            //
            // Link reference
            //
            if (typeof state.env.references === 'undefined') {
                return false;
            }

            // [foo]  [bar]
            //      ^^ optional whitespace (can include newlines)
            for (; pos < max; pos++) {
                const code = state.src.charCodeAt(pos);
                if (code !== 0x20 && code !== 0x0a) {
                    break;
                }
            }

            let label = '';

            if (pos < max && state.src.charCodeAt(pos) === 0x5b /* [ */) {
                const start = pos + 1;
                pos = md.helpers.parseLinkLabel(state, pos);
                if (pos >= 0) {
                    label = state.src.slice(start, pos++);
                } else {
                    pos = labelEnd + 1;
                }
            } else {
                pos = labelEnd + 1;
            }

            // covers label === '' and label === undefined
            // (collapsed reference link and shortcut reference link respectively)
            if (!label) {
                label = state.src.slice(labelStart, labelEnd);
            }

            const ref = state.env.references[md.utils.normalizeReference(label)];
            if (!ref) {
                state.pos = oldPos;
                return false;
            }
            href = ref.href;
            title = ref.title;
        }

        //
        // We found the end of the link, and know for a fact it's a valid link;
        // so all that's left to do is to call tokenizer.
        //
        if (!silent) {
            state.pos = labelStart;
            state.posMax = labelEnd;

            const tokens: Token[] = [];
            const newState = new state.md.inline.State(
                state.src.slice(labelStart, labelEnd),
                state.md,
                state.env,
                tokens,
            );
            newState.md.inline.tokenize(newState);

            const token = state.push('image', 'img', 0);
            token.children = tokens;
            token.attrs = [
                [ImsizeAttr.Src, href],
                [ImsizeAttr.Alt, ''],
            ];

            if (title) {
                token.attrs.push([ImsizeAttr.Title, title]);
            }

            if (width !== '') {
                token.attrs.push([ImsizeAttr.Width, width]);
            }

            if (height !== '') {
                token.attrs.push([ImsizeAttr.Height, height]);
            }

            if (opts?.enableInlineStyling) {
                let style: string | undefined = '';

                const widthWithPercent = width.includes('%');
                const heightWithPercent = height.includes('%');

                if (width !== '') {
                    const widthString = widthWithPercent ? width : `${width}px`;
                    style += `width: ${widthString};`;
                }

                if (height !== '') {
                    if (width !== '' && !heightWithPercent && !widthWithPercent) {
                        style += `aspect-ratio: ${width} / ${height};height: auto;`;
                        state.env.additionalOptionsCssWhiteList ??= {};
                        state.env.additionalOptionsCssWhiteList['aspect-ratio'] = true;
                    } else {
                        const heightString = heightWithPercent ? height : `${height}px`;
                        style += `height: ${heightString};`;
                    }
                }

                if (style) {
                    token.attrs.push([ImsizeAttr.Style, style]);
                }
            }
        }

        state.pos = pos;
        state.posMax = max;
        return true;
    };
};
