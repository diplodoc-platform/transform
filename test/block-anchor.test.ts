import dedent from 'dedent';

import initMarkdown from '../src/transform/md';
import plugin from '../src/transform/plugins/block-anchor';
import anchorsPlugin from '../src/transform/plugins/anchors';

const {parse, compile} = initMarkdown({plugins: [plugin, anchorsPlugin]});

describe('block-anchor', function () {
    it('renders block-anchor', () => {
        const input = '{%anchor my-anchor%}';
        const expected = '<hr id="my-anchor" class="visually-hidden" />';
        const actual = compile(parse(input));

        expect(actual).toBe(expected);
    });

    it('parses anchors surrounded by other blocks', () => {
        const input = '# Heading \n\n {%anchor my-anchor%} \n\n paragraph with content';
        const actual = compile(parse(input));
        const expected = dedent`
        <h1 id="heading">\
            <a href="#heading" class="yfm-anchor" aria-hidden="true">\
                <span class="visually-hidden">\
                    Heading\
                </span>\
            </a>\
            Heading\
        </h1>
        <hr id="my-anchor" class="visually-hidden" />\
        <p>paragraph with content</p>\n`;
        expect(actual).toEqual(expected);
    });

    it('parses block anchors', () => {
        const input = '${%anchor my-anchor} Content';
        const actual = compile(parse(input));
        const expected = '<p>${%anchor my-anchor} Content</p>\n';

        expect(actual).toEqual(expected);
    });

    it('works with heading anchors', () => {
        const input = '# Heading {#heading-anchor} \n {%anchor my-anchor%}';
        const expected = dedent`
        <h1 id="heading-anchor">\
            <a href="#heading-anchor" class="yfm-anchor" aria-hidden="true">\
                <span class="visually-hidden">\
                    Heading\
                </span>\
            </a>\
            Heading\
        </h1>
        <hr id="my-anchor" class="visually-hidden" />`;
        const actual = compile(parse(input));

        expect(actual).toBe(expected);
    });

    it('does not parse produce an anchor if there is content before markup', () => {
        const input = 'Content  {%anchor my-anchor%}';
        const expected = '<p>Content </p>\n';
        const actual = compile(parse(input));
        expect(actual).toBe(expected);
    });
});
