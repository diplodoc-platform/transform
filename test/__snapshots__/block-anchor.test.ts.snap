// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`block-anchor does not parse produce an anchor if there is content before markup 1`] = `
"<p>Content </p>
"
`;

exports[`block-anchor handles multiple anchors in the input 1`] = `
"<hr id=\\"first-anchor\\" class=\\"visually-hidden\\" /><p>Some content</p>
<hr id=\\"second-anchor\\" class=\\"visually-hidden\\" /><p>Some more content</p>
<hr id=\\"third-anchor\\" class=\\"visually-hidden\\" />"
`;

exports[`block-anchor parses anchors surrounded by other blocks 1`] = `
"<h1 id=\\"heading\\"><a href=\\"#heading\\" class=\\"yfm-anchor\\" aria-hidden=\\"true\\"><span class=\\"visually-hidden\\">Heading</span></a>Heading</h1>
<hr id=\\"my-anchor\\" class=\\"visually-hidden\\" /><p>paragraph with content</p>
"
`;

exports[`block-anchor parses block anchors 1`] = `
"<p>\${%anchor my-anchor} Content</p>
"
`;

exports[`block-anchor renders block-anchor 1`] = `"<hr id=\\"my-anchor\\" class=\\"visually-hidden\\" />"`;

exports[`block-anchor works with heading anchors 1`] = `
"<h1 id=\\"heading-anchor\\"><a href=\\"#heading-anchor\\" class=\\"yfm-anchor\\" aria-hidden=\\"true\\"><span class=\\"visually-hidden\\">Heading</span></a>Heading</h1>
<hr id=\\"my-anchor\\" class=\\"visually-hidden\\" />"
`;
